File: 74-search-a-2d-matrix/search-a-2d-matrix.py
class Solution(object):
    def searchMatrix(self, matrix, target):
        """
        :type matrix: List[List[int]]
        :type target: int
        :rtype: bool
        """
        
        rows = len(matrix)
        cols = len(matrix[0])
        count = 0

        left, right = 0, rows*cols-1

        while (left <= right):
            mid = (left + right) // 2

            mid_value = matrix[mid//cols][mid%cols]

            if mid_value == target:
                count += 1
                break

            elif mid_value > target:
                right = mid - 1

            else:
                left = mid + 1

        if count == 1:
            return True
        else:
            return False

File: 74-search-a-2d-matrix/search-a-2d-matrix.cpp
class Solution {
public:
    bool searchMatrix(vector<vector<int>>& matrix, int target) {
        int m = matrix.size();
        int n = matrix[0].size();

        int row = 0;

        while (row < m){
            int left = 0;
            int right = n-1;

            while (left <= right) {
                int mid = left + (right-left)/2;
                if (matrix[row][mid] < target) {
                    left = mid+1;
                }
                else if (matrix[row][mid] > target) {
                    right = mid-1;
                }
                else if (matrix[row][mid] == target){
                    return true;
                }
            }
            row++;
        }

        return false;
    }
};